sweep/sweeper.go

```diff
@@ -1,234 +1,627 @@
+	var sweepableInputs []Input
+	for _, input := range s.pendingInputs {
+		// Skip inputs that have a minimum publish height that is not
+		// yet reached.
+		if input.minPublishHeight > currentHeight {
+			continue
+		}
+
+		sweepableInputs = append(sweepableInputs, input.input)
+	}
+
+	return s.cfg.TxGenerator.Generate(sweepableInputs, satPerKW)
+}
+
+func (s *UtxoSweeper) sweep(inputs []Input,
+	satPerKW lnwallet.SatPerKWeight, currentHeight int32) error {
+
+	var err error
+
+	// Generate output script if no unused script available.
+	if s.currentOutputScript == nil {
+		s.currentOutputScript, err = s.cfg.GenSweepScript()
 		if err != nil {
 			return err
 		}
+	}
+
+	// Create sweep tx.
+	tx, err := s.cfg.TxGenerator.createSweepTx(inputs,
+		s.currentOutputScript,
+		uint32(currentHeight), satPerKW)
+	if err != nil {
+		return fmt.Errorf("cannot create sweep tx: %v", err)
+	}
 
-		sweepTx.TxIn[idx].Witness = witness
+	// Add tx before publication, so that we will always know that a
+	// spend by this tx is ours.
+	s.cfg.Store.NotifyPublishTx(tx)
 
-		return nil
+	// Publish sweep tx.
+	log.Debugf("Publishing sweep tx %v", tx.TxHash())
+	err = s.cfg.PublishTransaction(tx)
+
+	// Keep outputScript in case of an error, so that it can be
+	// reused for the next tx and causes no address inflation.
+	if err != nil {
+		log.Error(err)
+	} else {
+		s.currentOutputScript = nil
 	}
 
-	// Finally we'll attach a valid witness to each csv and cltv input
-	// within the sweeping transaction.
-	for i, input := range inputs {
-		if err := addWitness(i, input); err != nil {
-			return nil, err
+	// Reschedule sweep.
+	for _, input := range tx.TxIn {
+		pi := s.pendingInputs[input.PreviousOutPoint]
+
+		var nextAttemptDelta int32
+		switch err {
+		// Successful publish, but it may still not confirm.
+		// Retry later.
+		case nil:
+			nextAttemptDelta = 6
+
+		// In case of a double spend, retry the next block.
+		case lnwallet.ErrDoubleSpend:
+			nextAttemptDelta = 1
+
+		// For other errors, back off exponentially.
+		default:
+			nextAttemptDelta = 1 << uint(pi.publishAttempts)
 		}
-	}
 
-	return sweepTx, nil
+		pi.minPublishHeight = currentHeight + nextAttemptDelta
+		pi.publishAttempts++
+
+	}
+	return nil
 }
 
-// getWeightEstimate returns a weight estimate for the given inputs.
-// Additionally, it returns counts for the number of csv and cltv inputs.
-func (s *UtxoSweeper) getWeightEstimate(inputs []Input) ([]Input, int64, int, int) {
-	// We initialize a weight estimator so we can accurately asses the
-	// amount of fees we need to pay for this sweep transaction.
-	//
-	// TODO(roasbeef): can be more intelligent about buffering outputs to
-	// be more efficient on-chain.
-	var weightEstimate lnwallet.TxWeightEstimator
+// registerSpendNtfn registers a spend notification with the chain notifier. It
+// returns a cancel function that can be used to cancel the registration.
+func (s *UtxoSweeper) registerSpendNtfn(outpoint wire.OutPoint,
```

Comment:
I was a little confuse by this name, since it does more than the regular `registerforSpendNtfn` calls. Maybe `listenForSpend` or similar? :)

Reply:
renamed